### ISSUE TITLE
StackOverflowError in exception handling

### ISSUE THREAD

**Issue by erikogan** [jruby/jruby#4024]
### Environment

Similar errors in: 
- jruby 9.1.3.0-SNAPSHOT (2.3.0) 2016-07-20 8c97591 Java HotSpot(TM) 64-Bit Server VM 25.60-b23 on 1.8.0_60-b27 +jit [darwin-x86_64]
- jruby 9.1.2.0 (2.3.0) 2016-05-26 7357c8f Java HotSpot(TM) 64-Bit Server VM 25.60-b23 on 1.8.0_60-b27 +jit [darwin-x86_64]

Gems involved: 
- Rails 5.0.0
- RSpec 3.5.1
- aws-sdk 2.3.15-2.4.0 (2.3.14 does not throw the underlying exception)
### Expected Behavior

I expect a `TypeError: no implicit conversion of nil into String` to be raised due to a bug in aws-sdk.
### Actual Behavior

I am seeing a StackOverflowError exception when an RSpec integration test receives a `TypeError: no implicit conversion of nil into String`. 

Attempting to reproduce that error in isolation fails (the correct exception is raised)

Details here: https://github.com/aws/aws-sdk-ruby/issues/1232#issuecomment-234146274

When run with `JRUBY_OPTS=-X-C`, the following (truncated) stack trace occurs:

```
     Failure/Error: Unable to find org.jruby.ir.interpreter.InterpreterEngine.processBookKeepingOp(InterpreterEngine.java to read failed line
     Java::JavaLang::StackOverflowError:
     # org.jruby.ir.interpreter.InterpreterEngine.processBookKeepingOp(InterpreterEngine.java:368)
     # org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:196)
     # org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(Interpreter.java:132)
     # org.jruby.runtime.InterpretedIRBlockBody.callDirect(InterpretedIRBlockBody.java:102)
     # org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:64)
     # org.jruby.runtime.Block.call(Block.java:126)
     # org.jruby.RubyProc.call(RubyProc.java:342)
     # org.jruby.internal.runtime.methods.ProcMethod.call(ProcMethod.java:63)
     # org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:193)
     # org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:189)
     # org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:129)
     # org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:339)
     # org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:159)
     # org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:84)
     # org.jruby.internal.runtime.methods.InterpretedIRMethod.INTERPRET_METHOD(InterpretedIRMethod.java:157)
     # org.jruby.internal.runtime.methods.InterpretedIRMethod.call(InterpretedIRMethod.java:148)
     # org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:197)
     # org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:161)
     # org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:314)
     # org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:159)
     # org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:84)
     # org.jruby.internal.runtime.methods.InterpretedIRMethod.INTERPRET_METHOD(InterpretedIRMethod.java:157)
     # org.jruby.internal.runtime.methods.InterpretedIRMethod.call(InterpretedIRMethod.java:148)
[ Repeat another 143 times]
```

I will try to pare down the example to give a simple reproduction case. Unfortunately it’s in a full integration test, so there’s a lot going on.

**Comment by nightscape** [comment]
This might be related: https://github.com/jruby/jruby/pull/4117

**Comment by headius** [comment]
This is quite old and likely was fixed by my changes in #4117 (that is for pointing that out @nightscape). Closing as dupe.
